app/services/concerns/xccdf/profile_links.rb

```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+module Xccdf
+  # Methods related to saving rule references
+  module ProfileLinks
+    extend ActiveSupport::Concern
+
+    included do
+      def save_profile_links
+        ::ProfileLink.transaction do
+          ::ProfileLink.import!(
+            profile_links.select(&:new_record?), ignore: true
+          )
+
+          base = ::ProfileLink.joins(profile: :benchmark)
+                              .where('profiles.parent_profile_id' => nil)
+
+          profile_links_to_remove(base).delete_all
+        end
+      end
+
+      private
+
+      def profile_links
+        @profile_links ||= @op_profiles.flat_map do |op_profile|
+          profile_id = profile_id_for(ref_id: op_profile.id)
+          rule_and_rule_group_ids_for(ref_ids: op_profile.selected_rule_ids).map do |entity|
+            ::ProfileLink.find_or_initialize_by(
+              profile_id: profile_id, entity_id: entity[0], type: entity[1]
```

Comment:
Return objects, not arrays, activerecord can consume it:
```suggestion
              profile_id: profile_id, entity: entity
```

Reply:
Maybe this is where we're lacking the polymorphism? :thinking: I believe it should *Just Work &trade;*

Reply:
I wouldn't need to specify the type if I was doing `ProfileRule.find_or_initialize_by` or `ProfileRuleGroup.find_or_initialize_by` but since I'm doing `::ProfileLink.find_or_initialize_by` I needed to specify the type.. But I was thinking that this would be more efficient than doing ProfileRule and ProfileRuleGroup separately? Maybe?

Reply:
Oh also, when I only use `entity` instead of `entity_id` I get this error:
`ActiveRecord::StatementInvalid: PG::UndefinedColumn: ERROR:  column profile_links.entity does not exist`
But I know that it's usually possible to do this.. Maybe it isn't working correctly because of how to `belongs_to` is set up in ProfileRule and ProfileRuleGroup?